### ISSUE TITLE
Changing BaseThingHandler bridgeStatusChanged behaviour

### ISSUE THREAD

**Issue by J-N-K** [eclipse/smarthome#4074]
Currently this method works as follows:

bridgeStatus changes from ONLINE to OFFLINE: thingStatus is set to OFFLINE, thingStatusDetail to BRIDGE_OFFLINE
bridgeStatus changes from OFFLINE to ONLINE: checks whether the thingStatusDetail is BRIDGE_OFFLINE and if so, sets the thing Status to ONLINE

On my opinion this is wrong. If the thing was OFFLINE (e.g. because of mis-configuration) when the bridge goes offline, the thingStatusDetail (here: CONFIGURATION_ERROR) is lost, because it is set to BRIDGE_OFFLINE. When the bridge goes back online, the thingStatus changes to ONLINE, even if the reason for the thing to be OFFLINE in the first place is still there.

My suggestion is to change it so that the methods checks whether the thing is already OFFLINE and only set thingStatusDetail to BRIDGE_OFFLINE if the reason for the thing to go OFFLINE is indeed the bridge changing its status.

**Comment by maggu2810** [comment]
Just to summarize:
It is about the `bridgeStatusChanged` method implementation of `BaseThingHandler`.
Every binding can override this method and change the behavior to its need.

That being said, we can surely discuss the "base" implementation.
Do you think there is a rule that fits always? What should the user "see", is my thing offline because it is wrongly configured or because the bridge is offline?
* If the bridge is offline and the thing is correct configured the thing is offline => show bridge offline
* If the bridge is online and the thing is not correct configured the thing is offline => show configuration error
* If the bridge is offline and the thing is not correct configured the thing is offline => show ...

**Comment by maggu2810** [comment]
There are other transitions, too:
If the thing is still in unknown state (so, not online or offline), it is set to offline,bridge_offline, too and then to online if the bridge goes online...

**Comment by J-N-K** [comment]
That is exactly my point. The "base implementation" should check if the thingStatus was ONLINE before setting it to OFFLINE/BRIDGE_OFFLINE. And only set it ONLINE if it was OFFLINE/BRIDGE_OFFLINE (already implemented).

**Comment by triller-telekom** [comment]
I agree with @J-N-K. When the bridge goes OFFLINE, we should not blindly set the thing to OFFLINE too, but rather check if it is ONLINE. Because it is not ONLINE there usually is a reason for it which is set as a StatusDetail message which we should preserve.

So if the bridge goes OFFLINE, check if the attached thing is ONLINE or UNKNOWN, then set it to OFFLINE with reason "bridge offline" and otherwise keep its current OFFLINE state.

The only case which is unclear to me is, If the current OFFLINE state of the thing is without any reason while the bridge goes OFFLINE, should we leave it like that or in this specific case override the StatusDetail with "bridge OFFLINE"?

**Comment by davidgraeff** [comment]
I second this request. I always need to overwrite the method because the base implementation is not useful in many cases.

_Sent from my Lge LG-F700K using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by J-N-K** [comment]
@triller-telekom: If we set it to OFFLINE/BRIDGE_OFFLINE from UNKNOWN it is wrong to set it to ONLINE after the bridge goes to ONLINE. We don't know if that is true. We should set it to UNKNOWN then (or leave it at UNKNOWN when the bridge goes OFFLINE). Another possibility would be to set it to UNKNOWN/BRIDGE_OFFLINE (if that is allowed) and return to UNKNOWN/NONE if the bridge goes online again.

**Comment by maggu2810** [comment]
Isn't the whole problem what should be done if the bridge change to online again?
If a bridge changes to offline a thing should not be online anymore.
If a bridge changes to online it is binding specific if the things are online, still offline, ...

**Comment by kaikreuzer** [comment]
> If a bridge changes to offline a thing should not be online anymore.

I agree. The only valid status is OFFLINE.BRIDGE_OFFLINE in that situation.

> If a bridge changes to online it is binding specific if the things are online, still offline, ...

I agree as well. Contemplating about it, there is indeed no one-fits-all implementation. The current logic in the BaseThingHandler deals with the simple bindings, where the Things are accessible whenever the bridge is online (so they should rather be channels...). But @J-N-K is right that even in that simply situation, it could happen that a configuration is invalid and thus changing to ONLINE is not correct.

So how about only dealing with the switch to OFFLINE in the BaseThingHandler and require the individual bindings to deal with the comes-online case by themselves?

**Comment by triller-telekom** [comment]
> So how about only dealing with the switch to OFFLINE in the BaseThingHandler and require the individual bindings to deal with the comes-online case by themselves?

In this case I would prefer that a binding developer CANNOT override the "bridge goes offline" situation but is FORCED to override the "bridge goes online" situation. WDYT?

**Comment by kaikreuzer** [comment]
> a binding developer CANNOT override the "bridge goes offline" 

Well, besides setting the Thing status, he might want to take some other actions like resource cleanup or whatever. So we must allow him to be notified about that status change.

**Comment by maggu2810** [comment]
> So how about only dealing with the switch to OFFLINE in the BaseThingHandler and require the individual bindings to deal with the comes-online case by themselves?

I agree.
But isn't it a breaking change for all bindings, that don't handle bridge status changed itself?

**Comment by kaikreuzer** [comment]
Yes, it would be a breaking change - but from this discussion I would claim that any binding that does not yet override the method might anyhow have a buggy behavior that should be looked at more closely - this might be some work indeed :-/ Maybe worth to first investigate a bit the impact on the ESH bindings to get an idea.

**Comment by J-N-K** [comment]
@kaikreuzer While I agree that it is nearly impossible to find a general solution apllicable to all bindings for the transition BRIDGE_OFFLINE->BRIDGE_ONLINE, I still think that if the thing already is in OFFLINE state, this state should not be touched as it is very unlikely that this condition changes when the bridge goes online again.

What do you suggest? Splitting bridgeStatusChanged into bridgeStatusChangedToOffline and bridgeStatusChangedToOnline where the first has an implementation dealing with the most common situations and the latter is abstract so that the binding developer has to implement it by himself?

**Comment by kaikreuzer** [comment]
> I still think that if the thing already is in OFFLINE state, this state should not be touched as it is very unlikely that this condition changes when the bridge goes online again.

Well, you might be right that there is a good chance that the Thing might be offline for the same reason, but effectively, we do not know. When the bridge goes offline, the reason for the Thing to be offline is clearly the bridge - so setting it to OFFLINE.BRIDGE_OFFLINE is the only correct thing to do. Now if the bridge goes online again, we have to do a status transition to a correct status for the thing - we should not do guesswork here. So setting it to OFFLINE with a reason that it had (minutes? days? months?) before is as incorrect as setting it ONLINE. So either it should be set to UNKNOWN automatically or we leave it entirely to the binding to find the right current status.

> What do you suggest? Splitting bridgeStatusChanged into bridgeStatusChangedToOffline and bridgeStatusChangedToOnline

No, not really - this would be a breaking change and it probably does not make things easier for the implementors. My suggestion is as [mentioned above](https://github.com/eclipse/smarthome/issues/4074#issuecomment-323278745): "only dealing with the switch to OFFLINE in the BaseThingHandler and require the individual bindings to deal with the comes-online case by themselves".

**Comment by J-N-K** [comment]
I adjusted the PR accordingly: if the bridge goes offline -> thing OFFLINE/BRIDGe_OFFLINE, if bridge goes online -> thing UNKNOWN/NONE

**Comment by kaikreuzer** [comment]
"if bridge goes online -> thing UNKNOWN/NONE" -> hm, this wasn't part of my suggestion.
Having that in there means that everybody needs to implement both cases, while with my suggestion, people could simply call `super()` and only implement the "goes online" case....

**Comment by J-N-K** [comment]
Ok, but then every binding needs to be updated, otherwise the things will never go back online. I could prepare the changed for the ESH repo, adding methods that behave the same as the old basehandler.

**Comment by kaikreuzer** [comment]
Correct, that's what I stated [here](https://github.com/eclipse/smarthome/issues/4074#issuecomment-323281682).

So before going through this effort: @maggu2810 & @sjka, do you agree with it? Or do you have any better alternative?

**Comment by maggu2810** [comment]
If my reading has been correct, your suggestion https://github.com/eclipse/smarthome/issues/4074#issuecomment-324594265 fits exactly to what I think about.
If the bridge goes offline we could assume that the thing is offline, too. Every other transition is unclear.

**Comment by J-N-K** [comment]
With the latest discussions in #4086, what do you think about removing the implementation in BaseThingHandler (making it abstract) and leave the correct handling entirely to the binding?
